quinn/src/lib.rs

```diff
@@ -822,41 +807,50 @@ impl Future for Driver {
                             .pending
                             .entry(connection)
                             .or_insert_with(|| Pending::new(None));
-                        use quinn::Timer::*;
-                        let mut cancel = match timer {
-                            LossDetection => &mut pending.cancel_loss_detect,
-                            Idle => &mut pending.cancel_idle,
-                            Close => unreachable!(),
-                        };
                         let instant = endpoint.epoch + duration_micros(time);
-                        if let Some(cancel) = cancel.take() {
-                            let _ = cancel.send(());
-                        }
-                        let (send, recv) = oneshot::channel();
-                        *cancel = Some(send);
+                        match timer {
+                            Timer::LossDetection => {
+                                if let Some(mut key) = pending.cancel_loss_detect.as_ref() {
+                                    endpoint.timers.remove(key);
+                                }
+                                let key = endpoint
+                                    .timers
+                                    .insert_at((connection, Timer::LossDetection), instant);
+                                pending.cancel_loss_detect = Some(key)
+                            }
+                            Timer::Idle => {
+                                if let Some(mut key) = pending.cancel_idle.as_ref() {
+                                    endpoint.timers.remove(key);
+                                }
+                                let key = endpoint
+                                    .timers
+                                    .insert_at((connection, Timer::Idle), instant);
+                                pending.cancel_idle = Some(key);
+                            }
+                            Timer::Close => {
+                                endpoint
+                                    .timers
+                                    .insert_at((connection, Timer::Close), instant);
+                            }
+                        };
```

Comment:
Could this be made more concise and less redundant by refactoring it into two separate match expressions bracketing an `endpoint.timers.insert_at` call?

Reply:
I'm not sure what you mean. I've tried this, but that does not compile due the `Timer::Close` case:

```rust
let timer_key = match timer {
    Timer::LossDetection => {
        if let Some(mut key) = pending.loss_detect_timer.as_ref() {
            endpoint.timers.remove(key);
        }
        &mut pending.loss_detect_timer
    }
    Timer::Idle => {
        if let Some(mut key) = pending.idle_timer.as_ref() {
            endpoint.timers.remove(key);
        }
        &mut pending.idle_timer
    }
    Timer::Close => &mut None,  // cannot return a reference
};
*timer_key = Some(endpoint.timers.insert_at((connection, timer), instant));
```

Reply:
I think I got it. Does 31f61a0c1129cdb8dc7ccba49bc83c12167cc733 do what you described?